independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,176 @@
+        }
+
+        // Virtual threads needs to load the jarfile synchronously to avoid blocking. This means that eventually
+        // multiple threads could load the same jarfile in parallel and this duplication has to be reconciled
+        final var newJarFileRef = syncLoadAcquiredJarFile(jarResource);
+        if (jarResource.jarFileReference.compareAndSet(localJarFileRefFuture, newJarFileRef) ||
+                jarResource.jarFileReference.compareAndSet(null, newJarFileRef)) {
+            // The new file reference has been successfully published and can be used by the current and other threads
+            // The join() cannot be blocking here since the CompletableFuture has been created already completed
+            return consumeSharedJarFile(newJarFileRef.join(), jarResource, resource, fileConsumer);
+        }
+
+        // The newly created file reference hasn't been published, so it can be used exclusively by the current virtual thread
+        return consumeUnsharedJarFile(newJarFileRef, jarResource, resource, fileConsumer);
+    }
+
+    private static <T> T consumeSharedJarFile(JarFileReference jarFileReference,
+            JarResource jarResource, String resource, JarFileConsumer<T> fileConsumer) {
+        try {
+            return fileConsumer.apply(jarFileReference.jarFile, jarResource.jarPath, resource);
+        } finally {
+            jarFileReference.release(jarResource);
+        }
+    }
+
+    private static <T> T consumeUnsharedJarFile(CompletableFuture<JarFileReference> jarFileReferenceFuture,
+            JarResource jarResource, String resource, JarFileConsumer<T> fileConsumer) {
+        JarFileReference jarFileReference = jarFileReferenceFuture.join();
+        try {
+            return fileConsumer.apply(jarFileReference.jarFile, jarResource.jarPath, resource);
+        } finally {
+            boolean closed = jarFileReference.release(jarResource);
+            assert !closed;
+            // Check one last time if the file reference can be published and reused by other threads, otherwise close it
+            if (!jarResource.jarFileReference.compareAndSet(null, jarFileReferenceFuture)) {
+                closed = jarFileReference.release(jarResource);
+                assert closed;
+            }
+        }
+    }
+
+    private static CompletableFuture<JarFileReference> syncLoadAcquiredJarFile(JarResource jarResource) {
+        try {
+            return CompletableFuture.completedFuture(new JarFileReference(JarFiles.create(jarResource.jarPath.toFile())));
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to open " + jarResource.jarPath, e);
+        }
+    }
+
+    private static JarFileReference asyncLoadAcquiredJarFile(JarResource jarResource) {
+        CompletableFuture<JarFileReference> newJarRefFuture = new CompletableFuture<>();
+        CompletableFuture<JarFileReference> existingJarRefFuture = null;
```

Comment:
The introduction of this variable is the actual fix. The old algorithm [wrongly reused](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R168) the `newJarFileRef` one. This means that in the unfortunate case when a thread was [about to close the old jar](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R60), but [before setting the AtomicReference to null](https://github.com/quarkusio/quarkus/commit/ff4b32bb0d8a1df0ccf1cc86a1e1ae9a840c5b0f#diff-f20997e90aae1d2fe046b512defb1a3e14215b7a834177d43369542bae99a7f6R67), another thread could fail the `compareAndSet(null)` check here, take at the end of the loop from the `AtomicReference` the old and no longer valid `CompletableFuture` and in subsequent loop (after that the close finally managed to set the `AtomicReference` to null) mistakenly set it back into the same `AtomicReference`.